wandb/sdk/lib/filesystem.py

```diff
@@ -29,6 +31,60 @@ def mkdir_exists_ok(dir_name: StrPath) -> None:
         raise PermissionError(f"{dir_name!s} is not writable") from e
 
 
+def check_available_space(
+    path: StrPath,
+    size: Optional[int] = None,
+    reserve: int = 0,
+) -> int:
+    """Check that `path` has at least `size` + `reserve` available.
+
+    Args:
+        path: The path whose file system will be checked. It can be a file intended to
+            be written to, in which case `size` is optional.
+        reserve: The number of bytes to reserve on `path` in addition to the bytes
+            requested. [default: 0]
+        size: The number of bytes that will be written to `path`. If None, then
+            `path` must be a real file and its size will be used. [default: None]
+
+    Returns:
+        The number of bytes available, minus `reserve` and `size`.
+
+    Raises:
+        OSError [28]: if `path` has less than `size` available.
+        ValueError: if `size` is not None and `path` is not a real file.
+    """
+    path = Path(path).resolve()
+    if size is None:
+        if not path.is_file():
+            raise ValueError(
+                "If `size` is not specified then `path` must be a real file."
+            )
+        size = os.stat(path).st_size
```

Comment:
do you want to follow symlinks here?

Reply:
Oh that's very tricky.
No: if we follow symlinks then when we create an "adjacent" temp file we might be on a different file system than we checked for free space.
Yes: if we don't follow symlinks then we might end up on a different file system and mess up the assumption of atomic moves.

The `path` here on L62 can't be a symlink because the `.resolve()` call on L56 follows symlinks. But is that the correct behavior? The "expected" use of this function is to check paths that don't even exit yet to make sure there's enough space to write them. If you pass in a path that's a symlink to a real file on a different file system, what exactly does that mean? I think it means "I have this file here, I'm intending to copy it to an unspecified location", which is probably dangerous.

I think maybe the thing to do here is to name the `path` parameter `dest_path` and document it to make it clear this is the path you want to check for your ability to create it in the first place. Then remove the default size checking, and perhaps even check and prohibit calling with an extant path? Or perhaps just warn so that if multiple processes try to write to the same location none of them raise just because there's already a file there.

Reply:
ya sounds tricky, i was merely refering to the flag is `os.stat()` https://docs.python.org/3/library/os.html#os.stat , it defaults to True so wanted to make sure thats what you want